bots/tests-data

```diff
@@ -0,0 +1,254 @@
+SINKS = { }
+
+def main():
+    parser = argparse.ArgumentParser(description="Pull out test data for pull requests")
+    parser.add_argument("--open", action="store_true", help="Pull data on open pull requests")
+    parser.add_argument("--since", nargs="?", help="Since a given pull request or date")
+    parser.add_argument("-v", "--verbose", action="store_true", help="Show verbose progress")
+    opts = parser.parse_args()
+
+    for pull in pulls(opts.open and "open" or "closed", since=opts.since):
+        if opts.open:
+            merged = None
+        else:
+            merged = True
+        if opts.verbose:
+            sys.stderr.write("pull-{0}\n".format(pull["number"]))
+        for revision in revisions(pull):
+            if opts.verbose:
+                sys.stderr.write("- {0}\n".format(revision))
+            for (context, created, url, log) in logs(revision):
+                if opts.verbose:
+                    sys.stderr.write("  - {0} {1}\n".format(created, context))
+                for (status, name, body) in tap(log):
+                    sys.stdout.write(json.dumps({
+                        "pull": pull["number"],
+                        "revision": revision,
+                        "status": status,
+                        "context": context,
+                        "date": created,
+                        "head": merged,
+                        "test": name,
+                        "url": url,
+                        "log": body
+                    }))
+                    sys.stdout.write("\n")
+            if merged:
+                merged = False
+
+
+# An HTML parser that just pulls out all the <a href="...">
+# link hrefs in a given page of content. We also qualify these
+# hrefs with a base url, in case they're relative
+class HrefParser(HTMLParser.HTMLParser):
+    def __init__(self, base, hrefs):
+        HTMLParser.HTMLParser.__init__(self)
+        self.hrefs = hrefs
+        self.base = base
+
+    def handle_starttag(self, tag, attrs):
+        if tag.lower() == "a":
+            for (name, value) in attrs:
+                if name.lower() == "href":
+                    url = urlparse.urljoin(self.base, value)
+                    # print 'HREF', url
+                    self.hrefs.append(url)
+
+# Retrieves the content of the given URL
+def retrieve(url):
+    return urllib2.urlopen(url).read()
+
+# Returns a list of all results at the given URL
+def links(url):
+    result = [ ]
+    parser = HrefParser(url, result)
+    try:
+        parser.feed(retrieve(url))
+    except urllib2.HTTPError, ex:
+        if ex.code != 404:
+            raise
+    except urllib2.URLError:
+        pass
+    return result
+
+# Generates full GitHub JSON closed pull request
+# objects since the given date. We only care about closed
+# pull requests for processing, the rest are still in flux.
+def pulls(state='closed', since=None):
+    for pull in task.api.pulls(state=state, since=since):
+        yield pull
```

Comment:
You call this wrapper function exactly once (in `main()`) with explicit arguments, and this does not really "do" anything. Better just drop it and call `task.api.pulls()` directly in main for simplicity.

Reply:
Done.